### PR TITLE
ViTImageproc handle pil hw images

### LINKED ISSUE #34820: HW based PIL images not being handled in pretrained image processor
### System Info

transformers version 4.46.3

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
from datasets import load_dataset
import numpy as np

dataset = load_dataset("ylecun/mnist")
teacher_processor = AutoImageProcessor.from_pretrained("farleyknight-org-username/vit-base-mnist")

def process(examples):
    processed_inputs =  teacher_processor(examples["image"], input_data_format="none")`
    return processed_inputs
```

### Expected behavior

When using the mnist and dataset from `"ylecun/mnist"` and the finetuned VIT from `"farleyknight-org-username/vit-base-mnist"` from the same dataset using the loaded image processor fails, not allowing the shape of the image. Code is adapted from the [Knowledge Distillation for Computer Vision](https://huggingface.co/docs/transformers/main/en/tasks/knowledge_distillation_for_image_classification#knowledge-distillation-for-computer-vision)

The docstring for preprocess implies that this should be sufficient to allow for (height, width) format, but while resizing, there is only handling for ChannelDimension.First and ChannelDimension.Last, not for 'none' or ChannelDimension.None. 

the code fails under the calls stack

| Function | File | Line |
|:-----:| :----: | :----:|
| to_channel_dimension_format | image_transforms.py | 93
| to_pil_image | image_transforms.py | 204
| resize | image_transforms.py | 338
| resize | image_processing_vit.py | 138
| preprocess | image_processing_vit.py | 250

### PR DESCRIPTION
**Pull request by alem-147**
# What does this PR do?

fixes #34820 

Modifies image transformations and ViTImageProcessor such that HW PIL images can be have __call__ applied given the right arguments.

To supply an HW image to the ViTImageProcessor use the `input_data_format='none' or `input_data_format=ChannelDimension.NONE`. Further it is required that this is converted to RGB via the `do_convert_rgb=True` as the first option

```
    dataset = load_dataset("ylecun/mnist")
    processor = AutoImageProcessor.from_pretrained("farleyknight-org-username/vit-base-mnist")

    def process(examples):
        processed_inputs = processor(examples["image"],
                                             input_data_format="none",
                                             do_convert_rgb=True)
        return processed_inputs


    processed_dataset = dataset.map(process, batched=True)
``` 
or to leave it as a single channeled np.array by setting the mean and std of the image processor with `image_mean=[mean_value]` and `image_std=[std_value]`
```
...
    def process(examples):
        processed_inputs = teacher_processor(examples["image"],
                                             input_data_format="none",
                                             do_convert_rgb=False,
                                             image_mean=[.5],
                                             image_std=[.5])
        return processed_inputs
...
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
 @amyeroberts, @qubvel

### CODE REVIEW

File: src/transformers/image_transforms.py [hunk 1]
```diff
@@ -82,8 +82,11 @@ def to_channel_dimension_format(
 
     if input_channel_dim is None:
         input_channel_dim = infer_channel_dimension_format(image)
-
     target_channel_dim = ChannelDimension(channel_dim)
+
+    if target_channel_dim == ChannelDimension.NONE and input_channel_dim != ChannelDimension.NONE:
+        raise ValueError(f"Can't convert from dim format {input_channel_dim} to {target_channel_dim}")
+
```
Comment (qubvel) [inline src/transformers/image_transforms.py:89]: I would say we can expand extra dim, indead of raising an error here, what do you think?

File: src/transformers/image_transforms.py [hunk 2]
```diff
@@ -200,11 +203,12 @@ def to_pil_image(
...
-    # If the channel has been moved to first dim, we put it back at the end.
-    image = to_channel_dimension_format(image, ChannelDimension.LAST, input_data_format)
+    # If the channel has been moved to first dim, we put it back at the end if not 2d image
+    if input_data_format != ChannelDimension.NONE and image.shape != 2:
+        image = to_channel_dimension_format(image, ChannelDimension.LAST, input_data_format)
 
-    # If there is a single channel, we squeeze it, as otherwise PIL can't handle it.
-    image = np.squeeze(image, axis=-1) if image.shape[-1] == 1 else image
+        # If there is a single channel, we squeeze it, as otherwise PIL can't handle it.
+        image = np.squeeze(image, axis=-1) if image.shape[-1] == 1 else image
```
Comment (qubvel) [inline src/transformers/image_transforms.py:212]: Probably, we will not need this modification if `to_channel_dimension_format` will work with NONE format

File: src/transformers/image_transforms.py [hunk 3]
```diff
@@ -346,9 +350,11 @@ def resize(
...
         resized_image = np.expand_dims(resized_image, axis=-1) if resized_image.ndim == 2 else resized_image
         # The image is always in channels last format after converting from a PIL image
-        resized_image = to_channel_dimension_format(
-            resized_image, data_format, input_channel_dim=ChannelDimension.LAST
-        )
+        # if was initally a HW image then the added dim will be at the end already
+        if input_data_format != ChannelDimension.NONE:
+            resized_image = to_channel_dimension_format(
+                resized_image, data_format, input_channel_dim=ChannelDimension.LAST
+            )
```
Comment (qubvel) [inline src/transformers/image_transforms.py:357]: same here

File: src/transformers/image_transforms.py [hunk 4]
```diff
@@ -385,6 +391,11 @@ def normalize(
     if input_data_format is None:
         input_data_format = infer_channel_dimension_format(image)
 
+    # if of ChannelDimension.NONE and not made HWC by resize, handle here
+    if input_data_format == ChannelDimension.NONE:
+        input_data_format = ChannelDimension.LAST
+        image = np.expand_dims(image, axis=-1) if image.ndim == 2 else image
+
     channel_axis = get_channel_dimension_axis(image, input_data_format=input_data_format)
     num_channels = image.shape[channel_axis]
```
Comment (qubvel) [inline src/transformers/image_transforms.py:401]: Not sure se should expand dims here, can we handle without expand?

for example:
```python
if input_data_format != ChannelDimension.NONE:
    channel_axis = get_channel_dimension_axis(image, input_data_format=input_data_format)
    num_channels = image.shape[channel_axis]
else:
    num_channels = 1
```

File: src/transformers/image_utils.py [hunk 5]
```diff
@@ -239,6 +240,10 @@ def infer_channel_dimension_format(
         first_dim, last_dim = 0, 2
     elif image.ndim == 4:
         first_dim, last_dim = 1, 3
+    elif image.ndim == 2:
+        raise ValueError(
+            "must specify input_data_format = ChannelDimension.NONE and use supported image processor to use (height, width) image"
+        )
```
Comment (qubvel) [inline src/transformers/image_utils.py:246]: Why should we raise here instead of returning `ChannelDimension.NONE`?

File: src/transformers/models/vit/image_processing_vit.py [hunk 6]
```diff
@@ -242,7 +242,12 @@ def preprocess(
         )
 
         if do_convert_rgb:
-            images = [convert_to_rgb(image) for image in images]
+            if not input_data_format == ChannelDimension.NONE:
+                images = [convert_to_rgb(image) for image in images]
+            # handle HW PIL images to HWC by conversion to RGB - 3 channel allows for normalization with imagenet mean
+            else:
+                images = [convert_to_rgb(np.stack([image] * 3, axis=-1)) for image in images]
+                input_data_format = ChannelDimension.LAST
```
Comment (qubvel) [inline src/transformers/models/vit/image_processing_vit.py:251]: We should avoind doing this, insted we should extend `convert_to_rgb` function to handle NONE format

**Review by qubvel — COMMENTED** [review]
Hi @alem-147! Thanks for investigating the issue and proposing a PR! 
I left some comments regarding the code modification, let me know if you have additional questions. I would also appreciate adding tests for `NONE` format, thank you!
